### PR TITLE
Fix bots using player actor powers after defeat

### LINKED ISSUE #17456: AI is capable of using powers on player actor after being defeated.
This was the case in RV back then when we first added the Commander's Powers, we patched the issue out with a GrantConditionOnOwnerLost trait to disable the powers but the real issue behind it still stands.

### PR DESCRIPTION
**Pull request by reaperrr**
Fixes #17456.

### CODE REVIEW

**Comment by pchote** [comment]
We should really (also?) be blocking this in the `SupportPower` itself! Otherwise human players with a modified client may be able to exploit this.

**Comment by reaperrr** [comment]
Updated, moved the check to `SupportPowerManager` instead.

If its `Powers` list is empty, the module will automatically consider any remaining `waitingPowers` stale and clear them, too.

File: OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs [hunk 1]
```diff
@@ -94,6 +94,14 @@ void ActorRemoved(Actor a)
 
 		void ITick.Tick(Actor self)
 		{
+			if (Self.Owner.WinState != WinState.Undefined)
+			{
+				if (Powers.Count > 0)
+					Powers.Clear();
+
+				return;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs]: No need to return here - looping over an empty array is fine and lets us simplify down to
```csharp
// Dead players should never be able to use support powers
if (Self.Owner.WinState != WinState.Undefined && Powers.Count > 0)
	Powers.Clear();
```

**Comment by reaperrr** [comment]
Updated.

**Review by abcdefg30 — COMMENTED** [review]
Fix appears to work. :white_check_mark: 
However, wouldn't it be better to use `INotifyWinStateChanged.OnPlayerLost`? (We should use `OnPlayerWon` as well for completeness, although that shouldn't really be needed.)

**Comment by abcdefg30** [comment]
Never mind, a crash I found during testing I thought was a different issue is actually caused by this PR: We can't clear the powers of players that still have buildings, at least not without changing a few other locations like [SupportPowerChargeBar#L46](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Render/SupportPowerChargeBar.cs#L46). It might be enough to fix `GetPowersForActor` though.

Example crash log:
```
OpenRA engine version {DEV_VERSION}
Red Alert mod version {DEV_VERSION}
on map c03e7fe6cf205de831d90175512663ca4a03042f (East vs West by Chris Forbes).
Date: 2019-12-18 20:31:37Z
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.42000
Exception of type `System.Collections.Generic.KeyNotFoundException`: Der angegebene Schlüssel war nicht im Wörterbuch angegeben.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   at OpenRA.Mods.Common.Traits.Render.SupportPowerChargeBar.OpenRA.Traits.ISelectionBar.GetValue()
   at OpenRA.Mods.Common.Graphics.SelectionBarsAnnotationRenderable.DrawExtraBars(WorldRenderer wr, float2 start, float2 end)
   at OpenRA.Mods.Common.Graphics.SelectionBarsAnnotationRenderable.Render(WorldRenderer wr)
   at OpenRA.Graphics.WorldRenderer.DrawAnnotations()
   at OpenRA.Game.RenderTick()
   at OpenRA.Game.Loop()
   at OpenRA.Game.Run()
   at OpenRA.Game.InitializeAndRun(String[] args)
   at OpenRA.Program.Main(String[] args)
```
The crash happens here: https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs#L116

**Comment by abcdefg30** [comment]
Btw, thinking about it, I don't think we should always clear the powers, that will break the spectator overlay. Imho the best way to go forward would be to only clear if the game is not over and a player lost (i.e. check if the game is not over in `INotifyWinStateChanged.OnPlayerLost` and only then clear) since all other cases are irrelevant here.

**Comment by pchote** [comment]
Can you elaborate on the spectator overlay issue?

**Comment by abcdefg30** [comment]
I suppose you will not see any support powers although the buildings are still present, which is imho a bug. (It is not unusual to check the tab after a game ended.)

**Comment by pchote** [comment]
Why would the buildings still be present if they've lost?

**Comment by abcdefg30** [comment]
Usually not, but that can be the case if we're looking at a scripted mission. The overlay issue is mostly about players that won.

**Comment by pchote** [comment]
Ah, I see. I thought you meant "every tick" by always rather than "on win or loss".

**Comment by abcdefg30** [comment]
Sounds like an idea. Wouldn't that still hide the powers in the spectator overlay, though?

**Comment by pchote** [comment]
For players who have lost, yes. I don't see why that is a problem?

**Comment by abcdefg30** [comment]
It would look a bit unpolished for scripted maps I suppose, but no big problems.

**Comment by pchote** [comment]
The following works and should be pretty safe against regressions. Do you want to fix it here @reaperrr, or should I file a new PR?

```diff
diff --git a/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs b/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs
index 30e0909ae1..9704de9518 100644
--- a/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs
+++ b/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs
@@ -151,7 +151,7 @@ public class SupportPowerInstance
                protected int remainingSubTicks;
                public int RemainingTicks { get { return remainingSubTicks / 100; } }
                public bool Active { get; private set; }
-               public bool Disabled { get { return (!prereqsAvailable && !Manager.DevMode.AllTech) || !instancesEnabled || oneShotFired; } }
+               public bool Disabled { get { return Manager.Self.Owner.WinState != WinState.Undefined || (!prereqsAvailable && !Manager.DevMode.AllTech) || !instancesEnabled || oneShotFired; } }
 
                public SupportPowerInfo Info { get { return Instances.Select(i => i.Info).FirstOrDefault(); } }
                public bool Ready { get { return Active && RemainingTicks == 0; } }
```

**Comment by reaperrr** [comment]
Updated.

> Do you want to fix it here @reaperrr, or should I file a new PR?

Fixed it here.

**Comment by reaperrr** [comment]
> However, wouldn't it be better to use INotifyWinStateChanged.OnPlayerLost? (We should use OnPlayerWon as well for completeness, although that shouldn't really be needed.)

Hm, only saw this now. Need to check.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs [hunk 2]
```diff
@@ -109,7 +109,7 @@ public void ResolveOrder(Actor self, Order order)
 
 		public IEnumerable<SupportPowerInstance> GetPowersForActor(Actor a)
 		{
-			if (a.Owner != Self.Owner || !a.Info.HasTraitInfo<SupportPowerInfo>())
+			if (Powers.Count == 0 || a.Owner != Self.Owner || !a.Info.HasTraitInfo<SupportPowerInfo>())
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs]: Is this still needed with the `Disabled` fix below? I guess it can't hurt...

**Review by pchote — APPROVED** [review]
LGTM, and fix confirmed.

File: OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs (hunk 2 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs]: No it's not. It makes this even a bit more secure against crashes. (https://github.com/OpenRA/OpenRA/pull/17463#issuecomment-567207381 might have happened a different way.)

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/50b8c88a27c1fecb5512fe8b54564f8141da73ac...b7bddf19f7057ac09ea2ad48c969c9fbba623f53)

https://github.com/OpenRA/OpenRA/commit/a5b0b047ea320abef7d39ec84dcfa8ac75593efe

**Pull request merged**